### PR TITLE
[SMTChecker] Add support to array member length

### PR DESCRIPTION
**Pull request by leonardoalt**
Ref https://github.com/ethereum/solidity/issues/6048

Missing:
- [x] Inner arrays, as part of arrays, mappings and structs: `uint[][] arr; arr[0].length` etc
- [x] Arrays in different locations have a different type (storage vs memory). So either disregard location (easier) or add constraint that lengths are the same when copy storage -> memory / memory -> storage.
- [x] Assert that the lengths are still the same after an assignment

### CODE REVIEW

**Comment by leonardoalt** [comment]
Ready for review

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `3663baf99bd19e328b7e38543ba15b4ba05f431d`:
```
Coding style error:
 libsolidity/formal/CHC.h:33:#include "libsolidity/formal/SymbolicVariables.h"

```
Please check that your changes are working as intended.

File: libsolidity/formal/CHC.cpp [hunk 1]
```diff
@@ -654,6 +688,91 @@ vector<VariableDeclaration const*> CHC::stateVariablesIncludingInheritedAndPriva
 	return stateVars;
 }
 
+void CHC::createArrayLengths(Type const& _type)
+{
+	set<Type const*> types;
+	queue<Type const*> toVisit;
+	set<Type const*> visited;
+	toVisit.push(&_type);
+	while (!toVisit.empty())
```
Comment (leonardoalt) [inline libsolidity/formal/CHC.cpp]: Maybe use algorithms::BFS here

File: libsolidity/formal/CHC.cpp [hunk 2]
```diff
@@ -654,6 +688,91 @@ vector<VariableDeclaration const*> CHC::stateVariablesIncludingInheritedAndPriva
...
+void CHC::createFunctionArrayLengths(FunctionDefinition const& _function)
+{
+	for (auto const& var: _function.parameters())
+		createArrayLengths(*var->type());
+	for (auto const& var: _function.returnParameters())
+		createArrayLengths(*var->type());
+	for (auto const& var: _function.localVariables())
+		createArrayLengths(*var->type());
+
+	struct LengthCollector: public ASTConstVisitor
```
Comment (leonardoalt) [inline libsolidity/formal/CHC.cpp:744]: Any better ideas to write this visitor? I think our lambda based generic visitor wouldn't work here

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `02c483c385bbe1c611370176c8f1e22ffab6e071`:
```
Coding style error:
 libsolidity/formal/CHC.h:33:#include "libsolidity/formal/SymbolicVariables.h"

```
Please check that your changes are working as intended.

File: libsolidity/formal/CHC.cpp (hunk 2 above)
Comment (ekpyron) [reply libsolidity/formal/CHC.cpp:744]: Ref https://github.com/ethereum/solidity/issues/8315

File: libsolidity/formal/CHC.cpp (hunk 1 above)
Comment (ekpyron) [reply libsolidity/formal/CHC.cpp]: Yep, looks like a case for the algorithm.

Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp]: Yea, I was just waiting for you to mention it to check how to use it :p

**Comment by leonardoalt** [comment]
Updated with BFS

**Comment by leonardoalt** [comment]
The last commit seems to fix the solc-js smtchecker test failures.
Ready for review again @ekpyron

**Comment by leonardoalt** [comment]
This is probably going through with tuples instead

**Pull request closed without merging**